pkg/controller/statefulset/stateful_set.go

```diff
@@ -91,107 +78,106 @@ func NewStatefulSetController(podInformer cache.SharedIndexInformer, kubeClient
+		queue:      workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "statefulset"),
 	}
 
 	podInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
 		// lookup the statefulset and enqueue
-		AddFunc: psc.addPod,
+		AddFunc: ssc.addPod,
 		// lookup current and old statefulset if labels changed
-		UpdateFunc: psc.updatePod,
+		UpdateFunc: ssc.updatePod,
 		// lookup statefulset accounting for deletion tombstones
-		DeleteFunc: psc.deletePod,
+		DeleteFunc: ssc.deletePod,
 	})
-	psc.podStore.Indexer = podInformer.GetIndexer()
-	psc.podController = podInformer.GetController()
+	ssc.podStore.Indexer = podInformer.GetIndexer()
+	ssc.podController = podInformer.GetController()
 
-	psc.psStore.Store, psc.psController = cache.NewInformer(
+	ssc.setStore.Store, ssc.setController = cache.NewInformer(
 		&cache.ListWatch{
 			ListFunc: func(options metav1.ListOptions) (runtime.Object, error) {
-				return psc.kubeClient.Apps().StatefulSets(metav1.NamespaceAll).List(options)
+				return ssc.kubeClient.Apps().StatefulSets(v1.NamespaceAll).List(options)
 			},
 			WatchFunc: func(options metav1.ListOptions) (watch.Interface, error) {
-				return psc.kubeClient.Apps().StatefulSets(metav1.NamespaceAll).Watch(options)
+				return ssc.kubeClient.Apps().StatefulSets(v1.NamespaceAll).Watch(options)
 			},
 		},
 		&apps.StatefulSet{},
 		statefulSetResyncPeriod,
 		cache.ResourceEventHandlerFuncs{
-			AddFunc: psc.enqueueStatefulSet,
+			AddFunc: ssc.enqueueStatefulSet,
 			UpdateFunc: func(old, cur interface{}) {
 				oldPS := old.(*apps.StatefulSet)
 				curPS := cur.(*apps.StatefulSet)
 				if oldPS.Status.Replicas != curPS.Status.Replicas {
 					glog.V(4).Infof("Observed updated replica count for StatefulSet: %v, %d->%d", curPS.Name, oldPS.Status.Replicas, curPS.Status.Replicas)
 				}
-				psc.enqueueStatefulSet(cur)
+				ssc.enqueueStatefulSet(cur)
 			},
-			DeleteFunc: psc.enqueueStatefulSet,
+			DeleteFunc: ssc.enqueueStatefulSet,
 		},
 	)
 	// TODO: Watch volumes
-	psc.podStoreSynced = psc.podController.HasSynced
-	psc.syncHandler = psc.Sync
-	return psc
+	ssc.podStoreSynced = ssc.podController.HasSynced
+	return ssc
 }
 
 // Run runs the statefulset controller.
-func (psc *StatefulSetController) Run(workers int, stopCh <-chan struct{}) {
+func (ssc *StatefulSetController) Run(workers int, stopCh <-chan struct{}) {
 	defer utilruntime.HandleCrash()
 	glog.Infof("Starting statefulset controller")
-	go psc.podController.Run(stopCh)
-	go psc.psController.Run(stopCh)
+	if !cache.WaitForCacheSync(stopCh, ssc.podStoreSynced) {
+		return
+	}
+	go ssc.podController.Run(stopCh)
+	go ssc.setController.Run(stopCh)
 	for i := 0; i < workers; i++ {
-		go wait.Until(psc.worker, time.Second, stopCh)
+		go wait.Until(ssc.worker, time.Second, stopCh)
 	}
 	<-stopCh
 	glog.Infof("Shutting down statefulset controller")
-	psc.queue.ShutDown()
+	ssc.queue.ShutDown()
```

Comment:
The reason for deferring is that Shutdown will gracefully shutdown the queue and the workers will end up exiting gracefully in case this thread blows up before stopCh returns. It's a pattern we have started using in other controllers.